js/src/framework2.ts

```diff
@@ -44,6 +62,51 @@ export class Project {
     this.prompts = new PromptBuilder(this);
     this.scorers = new ScorerBuilder(this);
   }
+
+  public addPrompt(prompt: CodePrompt) {
+    this._publishablePrompts.push(prompt);
+    if (globalThis._lazy_load) {
+      globalThis._evals.prompts.push(prompt);
+    }
+  }
+
+  public addCodeFunction(
+    fn: CodeFunction<
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      any,
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      any,
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      GenericFunction<any, any>
+    >,
+  ) {
+    this._publishableCodeFunctions.push(fn);
+    if (globalThis._lazy_load) {
+      globalThis._evals.functions.push(fn);
+    }
+  }
+
+  async publish() {
+    await login();
+    const projectMap = new ProjectNameIdMap();
+    const functionDefinitions: FunctionEvent[] = [];
+    if (this._publishableCodeFunctions.length > 0) {
+      throw new Error(
+        "Code functions cannot be published directly. Use `braintrust push` instead.",
+      );
+    }
+    if (this._publishablePrompts.length > 0) {
+      for (const prompt of this._publishablePrompts) {
+        const functionDefinition =
+          await prompt.toFunctionDefinition(projectMap);
```

Comment:
curious why we couldn't do the functionDefinition within `addPrompt` is it a race condition of the `ProjectNameIdMap` may not have the latest information?

Reply:
just hygiene, so all of the network calling happens when you `publish()` rather than while defining.